### ISSUE TITLE
[New Concept Exercise]: Unpacking and Multiple Assignment in Python 

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3112]
This issue describes how to implement  the `unpacking` and `multiple assignment`  **concept exercise** for the Python track.
The related **concept documents issue** can be found [here](https://github.com/exercism/python/issues/3111).

<br>

## ✅ Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡 (<em>click to expand</em>)</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Stories](https://exercism.org/docs/building/tracks/stories)

## 🎯 Goal

This concept exercise is meant to teach an understanding/use of `unpacking` and the `*` (splat) and `**` (double splat) operators  in Python. 

<br>

## 💡Learning objectives

- Understand/use `unpacking` through the use of `*` and `**` _prefix_ operators in various scenarios
  - `*` and `**` as _prefixes_ ..... not to be confused with `*` (_multiply_) and `**` (_exponentiation_) as _infix_, or mathematical operators (**consider a link in the links doc or a mention in dig deeper.**)
  - ~~what happens in the process of "unpacking" - form, ordering, & iteration~~ (this goes in **dig deeper** or link docs.)
  - use in arguments to `functions`
  - use in argument _capture_ for `functions` (_aka passing an arbitrary number of arguments -- *args * & **kwargs_)
  - ~~use in defining `keyword only arguments`~~ (moved to argument exercise).
  - use in iterable (_mainly `tuple` and `list`_) unpacking & packing
  - use in `dict` unpacking & packing
- Understand/use `unpacking` via `multiple assignment`
  - using `multiple assignment ` in place of `indexing`
  - using `multiple assigment` + `*`  in place of `slicing`
  - ~~using "list-like" syntax & "tuple-like" syntax~~
  - unpacking plus "leftovers" via `*`
- Differences between straight  `multiple assignment` and `*` & `**`
- Deep unpacking

<br>

## 🤔 Concepts

- `unpacking`
- `unpacking generalizations`
- `multiple assignment`

<br>

## 🚫 Topics that are Out of scope

<br>
<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to expand</em>)</summary>
<br>

- `classes`
- `comprehensions`
- `comprehensions` in `lambdas`
- `map()`, `filter()` or `functools.reduce()` in a `comprehension`
-  `function-arguments` beyond explaining briefly how `*`, `**` work in function arguments, and how `*` works in requiring arguments.
- `functools` beyond `functools.reduce()`(_this will get its own exercise_)
- `generators`
- using an `assignment expression` or "walrus" operator (`:=`) alone or in a `lambda`

</details>
<br>

## ↩️  Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  
<br>

<details>
<summary>Prereqs  (<em>click to expand</em>)</summary>
<br>

- `basics`
- `bools`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `functional tools`
- Identity methods `is` and `is not`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>

##  📚 Resources for Writing and Reference

<details>
<summary><b>Resources</b> (<em>click to expand</em>)</summary>
<br>

- [Trey Hunner: Asterisks in Python - what they are and how to use them](https://treyhunner.com/2018/10/asterisks-in-python-what-they-are-and-how-to-use-them/)
- [Trey Hunner: Multiple assignment and tuple unpacking improve Python code readability](https://treyhunner.com/2018/03/tuple-unpacking-improves-python-code-readability/)
- [Dan Bader: Python Nested Unpacking](https://dbader.org/blog/python-nested-unpacking)
- [PEP 3132:  Extended Iterable Unpacking](https://www.python.org/dev/peps/pep-3132/)
- [PEP 0448: Additional Unpacking Generalizations](https://www.python.org/dev/peps/pep-0448/)
- [Python Tutorial:  Arbitrary Argument Lists](https://docs.python.org/3/tutorial/controlflow.html#arbitrary-argument-lists) & [Python Tutorial: Unpacking Argument Lists](https://docs.python.org/3/tutorial/controlflow.html#unpacking-argument-lists)
- [Python Docs: Expression Lists](https://docs.python.org/3/reference/expressions.html#expression-lists)
- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [built-ins: Python Docs](https://docs.python.org/3/library/functions.html)

</details>
<br>  

### Exercise Ideas & Stories

Should you need inspiration for an exercise story, you can find a collection [here](https://exercism.org/docs/building/tracks/stories).  You can also port an exercise from another track, **but please make sure to only to include tasks that actually make sense in Python** and that add value for a student.  Remove/replace/add tasks as needed to make the concept clear/workable.

<br>

##  📁 Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b> (<em>click to collapse</em>)</summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.

* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️ Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `a0d00a86-9eaa-487f-9114-e604263ccd96`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by meatball133** [comment]
This would be the first concept I would write but I like to give it a shot. I already have an exercise for this concept made for javascript(currently undergoing review). Which could be reused for python if we want and if it gets approved for javascript.

**Comment by BethanyG** [comment]
I would be delighted if you'd like to take this on!  Do you also want to do the concept write-up?  I'll assign both to you, and you can decide.  😄   Ping me if you have questions or want to talk through things/bounce ideas around.

**Comment by BethanyG** [comment]
PS 

> Which could be reused for python if we want and if it gets approved for javascript.

You don't have to wait for approval in JS if you think the same concept exercise works in a Python context.  I'd say go through and change the syntax and submit a draft here -- we can (quickly, I promise!) talk through it.  I don't have as firm a grip on JS as I do on Python, so can't enumerate off the top of my head where there would be differences -- but I'm sure they're not substantial.

**Comment by meatball133** [comment]
You can assign me to both. I will do a quick rewrite until tomorrow so I can show you the exercise in python

**Comment by meatball133** [comment]
The story(not 100% done):
# Instructions

Your friend is a train driver and has to drive cargo trains between cities.
Although your friend isn't amazing with handling the computers and would like some help with it.
Your friend would like your help organizing the train and correcting mistakes in the data.

```exercism/note
To practice, use the rest or spread operator to solve each of the tasks below.
```

## 1. Create a list of all wagons

Your friend has been keeping track of each wagon identifier. Although they are not sure how many wagons and would like the data to be returned as an array.

Implement a function `getListOfWagons` that accepts an unknown amount of positive integers that contains the id of each wagon.
It should return an array of all the wagon ids.

```javascript
getListOfWagons(1, 7, 12, 3, 14, 8, 5);
// => [1, 7, 12, 3, 14, 8, 3]
```

## 2. Move the first two elements to the end of the array

Now that you got a general feel for handling your friend's data.
The train id system works by the locomotive having id number one and the rest of the wagons having a random id.
Your friend had to connect two new wagons to the train and forgot to update the id system.
Now the first two wagons in the array have to be moved to the back of the train.
Your friend would like you to move the first two wagons to the end of the array.

Implement a function `fixListOfWagons` that accepts an array of the id of each wagon.
It returns an array where the 2 first elements are moved to the end of the array.

```javascript
eachWagonsWieght = [2, 5, 1, 7, 4, 12, 6, 3, 13];
fixListOfWagons(eachWagonsWieght);
// => [1, 7, 4,  12, 6, 3, 13, 2, 5]
```

## 3. Add missing values

Your friend realized that all data wasn't added and found another array which contains the missing wagons. 
Your friend would like you to add the missing values to the array. 
All they can remember is that the missing values should be placed after the locomotive in the array.

Given this new information, write a function called `CorrectListOfWagons` that takes two arrays which have the id of each wagon as an argument. 
The second array should be added after the first element of the first array.

```javascript
eachWagonsWieght = [1, 5, 20, 7, 4, 8];
missingWagons = [3, 17, 6, 15];
CorrectListOfWagons(eachWagonsWieght, missingWagons);
// => [1, 3, 17, 6, 15, 5, 20, 7, 4, 8]
```

## 4. Extend routing information

Now that the wagon data is correct, your friend would like you to update the routing information. 
Your friend has an object with the routing information and would like you to add more routing information to the object. 
Every route requires a bit different information so your friend would prefer a generic solution.

Implement a function `extendRouteInformation` that accepts two objects.
The first object contains which city the train should go between and the second object contains more routing information.
The function should return an object with the updated routing information.

```exercism/note
The variable `moreRouteInformation` can contain different properties.
```

```javascript
route  = {from: "Berlin", to: "Hamburg"};
moreRouteInformation = {length: "100", speed: "50"};
extendRouteInformation(route, moreRouteInformation);
// => {from: "Berlin", to: "Hamburg", length: "100", speed: "50"}
```

**Comment by BethanyG** [comment]
I actually think that's a great start!  I especially like the tasks that get at moving things in a list and also inserting them.  It's a little-discussed use case for multiple assignment -- but a powerful one, since you can replace many indexing and slicing scenarios with it.  The tests will be fun to write for that, since we may want to implement checks to push students away from alternate methods.

I also took a quick glance at the JS exercise, and think we could also adapt task 5.  I'll have a think on that one.

And as  I read through this and look at that laundry list of learning objectives...

<details>
<summary>Learning objectives</summary>

- Understand/use `unpacking` through the use of `*` and `**` _prefix_ operators in various scenarios
  - `*` and `**` as _prefixes_ ..... not to be confused with `*` (_multiply_) and `**` (_exponentiation_) as _infix_, or mathematical operators
  - what happens in the process of "unpacking" - form, ordering, & iteration
  - use in arguments to `functions`
  - use in argument _capture_ for `functions` (_aka passing an arbitrary number of arguments -- *args * & **kwargs_)
  - use in defining `keyword only arguments`
  - use in iterable (_mainly `tuple` and `list`_) unpacking & packing
  - use in `dict` unpacking & packing
- Understand/use `unpacking` via `multiple assignment`
  - using `multiple assignment ` in place of `indexing`
  - using `multiple assigment` + `*`  in place of `slicing`
  - using "list-like" syntax & "tuple-like" syntax
  - unpacking plus "leftovers" via `*`
- Differences between straight  `multiple assignment` and `*` & `**`
- Deep unpacking

</details>

I wonder if it make sense to cut this in half and do multiple assignment separate from unpacking?  I struggle, because the two are really intertwined.  But separation is an option, if you think it works better that way.   It could be within the same exercise theme, as a follow-on.

**Comment by BethanyG** [comment]
Ok.  Here are a few nits/thoughts  😄 

1.  We'll want to clean up the variable names at some point to be snake_case.  Not a big deal right now.
2.  Ditto for the references to arrays.  Python calls them `lists` 🙄 .  Again, so not a big deal right now. 
3.  Task 1 takes care of `*args`.  Wonder if we should insert a task 2 that is explicit about `**kwargs`? ( _As I side note, I think it's customary to name the parameters in this case `*args` & `**kwargs` to single them out as arbitrary length arguments, since there is also the `*` usage where you can otherwise unpack an iterable into a function call._)
4. About that unpacking an iterable into a function call.... we might want to do a task that requires that.  Although, that might not fit here either...
5. Deep unpacking.  [this article](https://treyhunner.com/2018/03/tuple-unpacking-improves-python-code-readability/) has some great examples.  Hard right now for me to think of a specific task that fits into this exercise, though.
6. I'm considering dumping the learning objective of using `*` for requiring keyword-only arguments.  We have a concept exercise all about arguments, so we could just leave it off.  Your thoughts?
7. The case where we _unpack_ and _repack_ `lists` and `tuples` - this is a great trick (which you use for `dicts` in task 4), especially with things like `enumerate` or when `zip` is used, and other scenarios.  Wonder if we can adapt the JS Task 5 to have a dictionary of `lists` or `tuples` That could then be used for `list`/`tuple` packing and unpacking??

**Comment by meatball133** [comment]
Well, the goal of the "train-driver" exercises for javascript was to be able to split 1 exercise into 2. The javascript exercise is made for the "rest and spread" operator. Which in python would be the `*` and `**`. There is another exercise that covers unpacking and deep unpacking for javascript. I more or less thought that it still could be a good starting point to build on. 

We could add an exercise that has `**kvargs` and then we could implement exercise 5 as well. We could also have an exercise that covers deep unpacking although after that point we will either have to merge or remove some of the already existing exercises. Since I think having more than 7 exercises in the same "file" is not a good idea.

**Comment by meatball133** [comment]
Also I didn't want to spend too much time rewriting the story to fit python if we didn't want to continue with this story

**Comment by meatball133** [comment]
I have worked on the syllabus. I see myself using "you" quite often. Is that something that should be avoided?

**Comment by meatball133** [comment]
Not 100% finished but I wanted to know if I missed something that should be in it or if I added something that shouldn't be in it.

# Unpacking and Multiple Assignment

Unpacking refers to the act of extracting the elements of a collection, such as a list, tuple, or dictionary, into variables.
With unpacking, there are some special operators used: `*` and `**`.
When unpacking a list or tuple, the `*` operator can be used to assign all the remaining elements to a variable.
When unpacking a dictionary, the `**` operator can be used to assign all the remaining key-value pairs to a variable.

When these operators are used without a collection it will pack the number of variables into a list, tuple, or dictionary.
It is common to use this kind of behavior when wanting functions that take a variable number of arguments.

Multiple assignment is the ability to assign multiple variables in one line.
This is done by separating the variables with a comma.

```exercism/note
`*<variable_name>` and `**<variable_name>` should not be confused with `*` and `**`. While `*` and `**` are used for multiplication and exponentiation respectively, `*<variable_name>` and `**<variable_name>` are used as packing and unpacking operators.
```

## Multiple assignment

Multiple assignment is the ability to assign multiple variables in one line.
This allows for code to be more concise and readable.
There has to be x number of variables on the left side of the `=` sign and x number of values on the right side of the `=` sign.
To separate the values, use a comma `,` and a space ` ` after each element.
Like following:

```python
>>> a, b = 1, 2
>>> a
1
```

Multiple assignment is not limited to 1 type of data type but can be used with any type of data type.
For example:

```python
>>> a, b, c = 1, "Hello", True
>>> a
1

>>> b
'Hello'

>>> c
True
```

Multiple assignment also allows swapping of variables in lists.
This practice is pretty common in linear sorting algorithms.
For example:

```python
>>> numbers = [1, 2]
>>> numbers[0], numbers[1] = numbers[1], numbers[0]
>>> numbers
[2, 1]
```

It is also possible to assign multiple valuables with the same value

```python
>>> a = b = 1
>>> a
1
>>> b
1
```

## Unpacking

In Python, it is possible to unpack a list/tuple/dictionary into distinct variables.
Since variables are mapped in lists with a specific order, it is therefore possible to unpack a list into variables in the same order.
Unpacking a list into variables:

```python
>>> fruits = ["apple", "banana", "cherry"]
>>> x, y, z = fruits
>>> x
"apple"
```

If there are values which are not needed then you can use `_` to ignore those values.

```python
>>> fruits = ["apple", "banana", "cherry"]
>>> _, _, z = fruits
>>> z
"cherry"
```

You can also do deep unpacking of lists, there you can unpack a list within a list.
Also known as a nested list.

```python
>>> fruits_vegatables = [["apple", "banana"], ["carrot", "potato"]]
>>> [[a, b], [c, d]] = fruits_vegatables
>>> a
"apple"

>>> d
"potato"
```

We can also mix using deep unpacking and normal unpacking.

```python
>>> fruits_vegatables = [["apple", "banana"], ["carrot", "potato"]]
>>> [a, [c, d]] = fruits_vegatables
>>> a
["apple", "banana"]

>>> c
"carrot"
```

If the unpacking has variables with incorrect placement and/or incorrect number of values then you will get a `ValueError`.

```python
>>> fruits_vegatables = [["apple", "banana"], ["carrot", "potato"]]
>>> [[a, b], [d]] = fruits_vegatables

ValueError: too many values to unpack (expected 1)
```

### Unpacking a list/tuple with `*`

When unpacking a list/tuple you can use the `*` operator to get the rest of a list/tuple.
We can for example get the first element and then get the rest into a new list without the first element.

```python
>>> fruits = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
>>> x, *last = fruits
>>> x
"apple"

>>> last
["banana", "cherry", "orange", "kiwi", "melon", "mango"]
```

We can also get variables at the end of the array and in the front of a list.

```python
>>> fruits = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
>>> x, *middle, y, z = fruits
>>> y
"melon"

>>> middle
["banana", "cherry", "orange", "kiwi"]
```

We can also use `*` for deep unpacking.

```python
>>> fruits_vegatables = [["apple", "banana", "melon"], ["carrot", "potato", "tomato"]]
>>> [[a, *rest], b] = fruits_vegatables
>>> a
"apple"

>>> rest
["banana", "melon"]
```

### Unpacking a dictionary

Unpacking a dictionary is a bit different than unpacking a list/tuple.
When unpacking a dictionary you can only unpack the keys and not the values.

```python
>>> fruits_inventory = {"apple": 6, "banana": 2, "cherry": 3}
>>> x, y, z = fruits_inventory
>>> x
"apple"
```

If you want to unpack the values then you can use the `values()` method.

```python
>>> fruits_inventory = {"apple": 6, "banana": 2, "cherry": 3}
>>> x, y, z = fruits_inventory.values()
>>> x
6
```

You can also unpack the keys and values into two different variables.
Using `items()` will generate tuples with key-value pairs.

```python
>>> fruits_inventory = {"apple": 6, "banana": 2, "cherry": 3}
>>> x, y, z = fruits_inventory.items()
>>> x
("apple", 6)
```

## Packing

Packing is the ability to pack multiple variables into one variable.
This is done by using the `*` or `**` operator.
This is useful when you want to unpack and do changes and then pack it back into a variable.
It also makes it possible to merge 2 or more lists/tuples/dictionaries.

### Packing a list/tuple with `*`

Packing a list/tuple is done by using the `*` operator.
This will pack all the variables into a list/tuple.

```python
>>> fruits = ["apple", "banana", "cherry"]
>>> more_fruits = ["orange", "kiwi", "melon", "mango"]
>>> combined_fruits_lists = [*fruits, *more_fruits]

>>> combined_fruits_lists
["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
```

### Packing a dictionary with `**`

Packing a dictionary is done by using the `**` operator.
This will pack all the variables into a dictionary.

```python
>>> fruits_inventory = {"apple": 6, "banana": 2, "cherry": 3}
>>> more_fruits_inventory = {"orange": 4, "kiwi": 1, "melon": 2, "mango": 3}
>>> combined_fruits_inventory = {**fruits_inventory, **more_fruits_inventory}

>>> combined_fruits_inventory
{"apple": 6, "banana": 2, "cherry": 3, "orange": 4, "kiwi": 1, "melon": 2, "mango": 3}
```

## Usage of `*` and `**` with a function

### Packing with a function

When you have a function with an unknown amount of arguments or a large number of arguments then you can use `*args` or `**kwargs` to pack the arguments.
`*args` is used for packing a non-keyworded variable length argument list.
`**kwargs` is used for packing keyworded variable length of arguments to a function.

Usage of `*args`:

```python
>>> def my_function(*args):
...     print(args)

>>> my_function(1, 2, 3)
(1, 2, 3)

>>> my_function("Hello", "World")
("Hello", "World")
```

Usage of `**kwargs`:

```python
>>> def my_function(**kwargs):
...   print(kwargs)

>>> my_function(a=1, b=2, c=3)
{"a": 1, "b": 2, "c": 3}
```

`*args` and `**kwargs` can also be used in combination with each other.
This allows splitting the keyworded and non-keyworded arguments.

```python
>>> def my_function(*args, **kwargs):
...   print(args)
...   print(kwargs)

>>> my_function(1, 2, 3, a=1, b=2, c=3)
(1, 2, 3)
{"a": 1, "b": 2, "c": 3}
```

You can also write arguments infront of `*args` and `**kwargs` to allow for normal arguments.

```exercism/note
While you can have arguments in front so can't you have after since that will raise an error.
Also, you can't combine normal arguments with both `*args` and `**kwargs`.
```

```python
>>> def my_function(a, b, *args):
...   print(a)
...   print(b)
...   print(args)

>>> my_function(1, 2, 3, 4, 5)
1
2
(3, 4, 5)
```

### Unpacking with functions

When having a list or a tuple of arguments you can use `*` to unpack the list/tuple into arguments for a function.

```python
>>> def my_function(a, b, c):
...   print(a)
...   print(b)

numbers = [1, 2, 3]
>>> my_function(*numbers)
1
2
```

**Comment by meatball133** [comment]
Not 100% finished but I wanted to know if I missed something that should be in it or if I added something that shouldn't be in it.

**Comment by BethanyG** [comment]
> I have worked on the syllabus. I see myself using "you" quite often. Is that something that should be avoided?

I don't think it's bad.  It is conversational and familiar -- but so is the exercise setup.  It talks about a friend coming to the student for some problem solving, and is written second-person. 

I did do a re-write, to see if you liked it better.  But since we are both online, do you want to hop in a call and pair on stuff?

**Comment by meatball133** [comment]
Sure

**Comment by BethanyG** [comment]
Ok.  Let me transfer to a desk & get  a link.  😄

**Comment by BethanyG** [comment]
Are you on Slack?  If so I will DM there.

**Comment by BethanyG** [comment]
> The story(not 100% done):
>
> # Instructions
>
> Your friend is a Locomotive Engineer who drives cargo trains between cities.
> Although your friend is great handling the trains, they aren't amazing handling the logistics computers and would like your  programming help organizing the train and correcting mistakes in the data.
>
> ```exercism/note
> To practice, use the Unpacking and Multiple Assignment to solve each of the tasks below.
> ```
>
> ## 1. Create a list of all wagons
>
> Your friend has been keeping track of each wagon identifier, but they're never sure how many wagons they are going to have to process at any given time. It would be much easier for the rest of the logistics program to have the data to be returned as a list.
>
> Implement a function `getListOfWagons` that accepts an unknown amount of positive integers which are the IDs of each wagon.
> It should then return the given IDs as a `list` .
>
> ```javascript
> getListOfWagons(1, 7, 12, 3, 14, 8, 5);
> // => [1, 7, 12, 3, 14, 8, 3]
> ```
>
> ## 2. Move the first two elements to the end of the array
>
> At this point,  you are starting to get a feel for your friend's data and how it's used in the logistics program.
> The train ID system works by assigning the locomotive an ID of `1` and then assigning the remainder of the wagons a randomly chosen ID greater than `1`.
>
> But then your friend had to connect two new wagons to the train and forgot to update the system!
> Now the first two wagons in the `list` have to be moved to the back of the train, or everything will be out of order.
> Your friend would be really grateful to you for fixing their mistake.
>
> Implement a function `fixListOfWagons` that accepts a wagon ID  `list` .
> The function should `return` the `list`  with the first 2 elements repositioned to the end of the `list`.
>
> ```javascript
> eachWagonsWieght = [2, 5, 1, 7, 4, 12, 6, 3, 13];
> fixListOfWagons(eachWagonsWieght);
> // => [1, 7, 4,  12, 6, 3, 13, 2, 5]
> ```
>
> ## 3. Add missing values
>
> Uh-oh.  some wagons seem to have gone missing.
>
> Fortunately, your friend just found another `list` which appears to contain the missing wagon IDs, and would like you to add them into the main wagon ID `list`. 
> All they can remember is that the missing values should be placed _directly after_ the designated locomotive.
>
> Given this new information, write a function called `CorrectListOfWagons` that takes two `lists` containing wagon IDs as the arguments. 
> The wagon IDs of the second `list` should be added into the first `list` directly after the locomotive (ID `1`).
>
> ```javascript
> eachWagonsWieght = [1, 5, 20, 7, 4, 8];
> missingWagons = [3, 17, 6, 15];
> CorrectListOfWagons(eachWagonsWieght, missingWagons);
> // => [1, 3, 17, 6, 15, 5, 20, 7, 4, 8]
> ```
>
> ## 4. Extend routing information
>
> Now that all the wagon data is correct, your friend would like you to update the systems routing information. 
> Initial routing information has been constructed as a `dict` , and you friend would like you to update it with the additions provided. 
> Every route requires slightly different information, so your friend would really prefer a generic solution.
>
> Implement a function `extendRouteInformation` that accepts two `dicts`.
> The first `dict` contains which cities the train route moves between.
>
> The second `dict` contains other routing details such as train speed or length.
> The function should return a consolidated `dict` with all routing information.
>
> ```exercism/note
> The variable `moreRouteInformation` can contain different properties.
> ```
>
> ```javascript
> route  = {from: "Berlin", to: "Hamburg"};
> moreRouteInformation = {length: "100", speed: "50"};
> extendRouteInformation(route, moreRouteInformation);
> // => {from: "Berlin", to: "Hamburg", length: "100", speed: "50"}
> ```

**Comment by meatball133** [comment]
Alright, some changes:

Added links (feel free to change them/ add some).

Took in your feedback, I am not sure I covered everything so if you find something just let me know.
